File: benchmarks/linux/ipc_bench.c
#include <stdio.h>
#include <string.h>
#include <time.h>
#include <unistd.h>

#define ITERATIONS 1000
#define MSG_SIZE 1024

int main() {
    struct timespec start, end;
    double total = 0.0;
    char sendBuf[MSG_SIZE], recvBuf[MSG_SIZE];
    memset(sendBuf, 'X', MSG_SIZE);

    for (int i = 0; i < ITERATIONS; i++) {
        int pipefd[2];
        if (pipe(pipefd) < 0) { perror("pipe"); return 1; }

        clock_gettime(CLOCK_MONOTONIC, &start);

        write(pipefd[1], sendBuf, MSG_SIZE);
        read(pipefd[0], recvBuf, MSG_SIZE);

        clock_gettime(CLOCK_MONOTONIC, &end);
        double elapsed = (end.tv_sec - start.tv_sec) * 1000.0
                       + (end.tv_nsec - start.tv_nsec) / 1e6;
        total += elapsed;

        close(pipefd[0]);
        close(pipefd[1]);
    }

    printf("RESULT:%.4f\n", total / ITERATIONS);
    return 0;
}

File: benchmarks/linux/ipc_bench.py
"""Linux IPC benchmark: pipe() / write() / read()."""
import os
import time

ITERATIONS = 1000
MSG        = b'X' * 1024
total      = 0.0

for _ in range(ITERATIONS):
    r, w = os.pipe()

    t0 = time.perf_counter()
    os.write(w, MSG)
    os.read(r, 1024)
    t1 = time.perf_counter()

    total += t1 - t0
    os.close(r)
    os.close(w)

print(f'RESULT:{(total / ITERATIONS) * 1000:.4f}')
